### ISSUE TITLE
pthreads not supported in apache

### ISSUE THREAD

**Issue by palmtown** [krakjoe/pthreads#598]
Shouldn't the ability to enable/disable pthreads in apache be up to the user who's environment is at stake?  Please enable pthreads with apache support off by default and allow the user to turn it on.

**Comment by palmtown** [comment]
No comment?

**Comment by krakjoe** [comment]
I didn't disable pthreads in web server environments to annoy you; I done it to keep you safe, and will not reverse that decision.

PHP and Apache do not provide the API required to make it *properly* safe.

If I patched Apache and PHP tomorrow to include some additional hooks in an effort to make it properly safe, you would *never* be able to make it scale. 

Threads are not a utility for an Apache worker mpm server, they are part of the architecture of the software; Coming in and starting additional threads inside that architecture is reckless and destroys the ability of that architecture to scale and operate as it was intended.

It never made sense to use pthreads inside Apache, I left it working in previous versions in an attempt to lower the entry barrier (everyone is comfortable with a web server). That was a mistake, my mistake; Threading is not simple, when you try to make something inherently complicated simple, you just make it stupid. 

Allowing threads to be started inside Apache without the ability to make either PHP or Apache properly aware of the change in architecture was abusive, and extremely stupid.

**Comment by palmtown** [comment]
Hello there,

First and foremost, thank you for your comments.  Secondly, while I agree that you 100% that we should have a legitimate concern with safety, I disagree with you that there is one right answer and that's disablement.  

Third, I believe your decision to disable function is apache is one part concern, and one part a bad moral decision.  Why?  Let me elaborate by giving you an analogy.  History has proven that Guns in the wrong hands kill people, but does that mean you take them away from everyone?  No, because that would be an amoral decision and would do an injustice to the people who has handled them with care and safety.

Let me give you another analogy, Soda is bad for you and it increases your chances of diabetes, risk of heart attack etc., but does that mean you put a ban on soda?  No, because again, that would do an injustice to those who drink it in portions that doesn't contribute to these health risks.

Therefore, disabling pthreads in apache is no different from banning soda and/or guns, it's amoral.  

In conclusion, this is larger than you or I and it has to be a decision for the stakeholder to make.  What happens on XYZ server is not your liability, however, the liability of the people who decides to enable it.

Having stated that, I would ask that you enable it and make a moral decision restoring choice, not force by ignorance.

**Comment by smolinari** [comment]
Amoral? That is the only argument you can offer? How about a decent technical reason why you absolutely need pthreads working with Apache? 

Scott

**Comment by palmtown** [comment]
Evolution? Steve Jobs and Steve Wozniak changed the world forever as we know it because they evolved their ideas.  

With respect to "why you absolutely need pthreads working with Apache?" because I want to evolve my application beyond what it is today.  Because the one thing set this project apart from anything like it was taken away without due process.  

You see, there's many ways to implement multi-threading in the command line without pthreads, however, there's little to none great that can do it via LAMP. 

In conclusion, it has to be restored so that it can evolve for the successful life of the project--its bigger than you or I.

**Comment by smolinari** [comment]
Sorry, but allowing PHP to thread in Apache is not evolution. It's a train wreck waiting to happen. Any project being built on threading in Apache is being built on a technology that simply can't work well (which is Joe's message above too) and thus is pretty much doomed to failure. You should be thanking Joe that he is saving you from this peril. 

If you want to have a proper and "Apache-like" threaded web server system with PHP, have a look at appserver.io. Your project can probably also profit from their hard work. Have a look! :smile: 

Scott

**Comment by palmtown** [comment]
Scott, I've taken a look at appserver.io, my first time hearing about it--thanks for the recommendation.  The message queue service seems comparable, however, there are some general concerns.

The main issue is that its much easier, efficient, scalable and safer to evolve pthreads opposed to a new technology that has a lot of overhead and seems to lag behind in compatibility.  

I understand the technological challenge that exists in pthreads, but for it to get passed those limitations it must be open.  There's nothing learned from a perfect world, you have to be given the opportunity to fail to evolve.

Having stated that, I'll continue my search to find the right solution for me, nonetheless, I still disagree that the option should be extended and a reach should be backed to evolve, not destroy.

**Comment by smolinari** [comment]
Um, evolution actually doesn't come from failure. It comes from success. Survival of the fittest and all. 

Joe warns clearly that in a regular web server stack, threading has no real purpose. As Joe put it...

> you can't use pthreads safely and sensibly anywhere but CLI.

I would imagine the only thing that can probably get Joe's attention is a very good viable use case for threading in a standard web server, which you fail to explain in any kind of detail. 

It seems you are looking for or are creating a "message queue". That is all I can gather out of our conversation. There are a number of solutions for asynchronous work with and without threading in PHP and also with external systems. What is it you are looking for exactly? 

Also, if I may ask. Why Apache? Apache is on its way out. Nginx is the better, more accepted solution these days. 

Scott

**Comment by palmtown** [comment]
Hey Scott,

Great dialogue, however, I would have to disagree with you that evolution doesn't involve failure.  You can name any great inventor, and I'd be willing to bet the Dow Jones they've had some form of failure that lead to success.  As scientist/inventors/developers we test things that often fail to arrive at a successful conclusion.

In terms of Joe's warning, again, he's right, 100%.  But that doesn't mean you protect by forceful removal of rights to all.  

In terms of apache/nginx, Apache still beats Nginx when using it solely for PHP, based on the blogs I've read.  For example, http://www.eschrade.com/page/why-is-fastcgi-w-nginx-so-much-faster-than-apache-w-mod_php/.

Having stated that, if the option was enabled to fail, maybe I would be willing to invest resources to help make it better.  Actually, maybe the many people that has asked for the same thing would of been willing to do the same.  And through that, we may have arrived at a better place, even another spin off of pthreads.

However, since the opportunity was taken away, the chances of it happening are even more limited.  Failure is what makes success possible.  In other words, you can't feel good about something till you know what bad feels like.

Therefore, I'll agree to disagree with you, however, in basic moral principals, you should not take away from one group, to solve an issue with another group and that is what have been done.  That's my argument.

In terms of what I'm looking for.  I appreciate your recommendation; but what I'm looking for is a true PHP multi-threader.  When my app gets one request, I have 15 or more things I need to do that the user can not wait for.  There are many ways to achieve it, by far pthreads was the best given it was allowed in Nginx or Apache.  But since it can not be, I'm better off just doing what we've been doing for years, which is to spend 200 to 300 ms doing a non-blocking curl post with my request.

**Comment by smolinari** [comment]
I guess you meant innovation and not evolution then and thus, I actually agree. It takes a good number of attempts, before any innovation becomes successful. 

As for your project, I'd still bet Joe will need more than a questioning of his judgement
 to motivate him to change his position. You'll need to bring up that use case and clear technical need.

You also have these possibilities:

https://github.com/icicleio/icicle
https://github.com/reactphp/react

Scott
